In [1]:
import pandas as pd

/Users/admin/anaconda/lib/python3.6/site-packages/matplotlib/__init__.py:913: UserWarning: axes.color_cycle is deprecated and replaced with axes.prop_cycle; please use the latter.
  warnings.warn(self.msg_depr % (key, alt_key))


# size of data

There are 518077 records and 76 variables(including target variable "EV_LARGE"

Column 41 has mixed data type.


In [2]:
# reading csv claimdata_utf8
lloss = pd.read_csv('/Users/admin/Desktop/AXAproject/claimdata_utf8.csv', low_memory=False)
lloss.shape

(518077, 76)

# Brief review of dataset

In [3]:
lloss.head()

Event_Number  Claim_Number Event_Occurrence_Month Inception_Month  \
0       7541881           NaN               2011/1/1        2010/4/1   
1       7541914     8981138.0               2011/1/1       2010/12/1   
2       7541914     8981144.0               2011/1/1       2010/12/1   
3       7541936     8960750.0               2011/1/1        2010/2/1   
4       7541986     8960833.0               2011/1/1        2010/6/1   

  Issued_Date Issued_Time  Coverage1  Coverage2  PI_Class  IN_OUTPATIENT  \
0   2010/4/24       14:23        NaN        NaN       NaN            NaN   
1  2010/11/11       16:37        6.0        6.0       3.0            1.0   
2  2010/11/11       16:37        3.0        3.0       3.0            1.0   
3    2010/2/8       14:32        2.0        2.0       NaN            NaN   
4    2010/6/4       14:26        2.0        2.0       NaN            NaN   

       ...        Car_Age Previous_Accidents Insured_Sex  Insured_Age  \
0      ...             12                  0           2           36   
1      ...              0                  0           1           61   
2      ...              0                  0           1           61   
3      ...             13                  0           1           48   
4      ...              7                  0           1           67   

  Insured_Age_Range   Region  Prefecture  Prev_Insurer_Category  \
0             30-39  (02) 東北         福島県        (02) Non Direct   
1               60-  (07) 九州         福岡県        (02) Non Direct   
2               60-  (07) 九州         福岡県        (02) Non Direct   
3             40-49  (07) 九州         熊本県               (99) New   
4               60-  (03) 関東         茨城県        (02) Non Direct   

   Original_Channel  Renewal_Times  
0        (01) Phone              7  
1        (01) Phone              7  
2        (01) Phone              7  
3        (01) Phone              1  
4       (03) Broker              1  

[5 rows x 76 columns]

#we found a few variables didn't appear on this dataset but were described in the item list
- "Paid"
- "OS"
- "incurred_Band"
- "Patient_Class"
- "Bonus_Malus_Band"
- "Insured_Age_Band"
- "Prev_Insured"

#we found some variables did appear on the dataset but weren't included in the item list
- "IN_OUTPATIENT"
- "Bonus_Malus_Range"
- "Insured_Age_Range"

In [4]:
lloss.columns

Index(['Event_Number', 'Claim_Number', 'Event_Occurrence_Month',
       'Inception_Month', 'Issued_Date', 'Issued_Time', 'Coverage1',
       'Coverage2', 'PI_Class', 'IN_OUTPATIENT', 'Incurred', 'EV_LARGE',
       'Driver', 'Contract_Status', 'Reason', 'Effective_Date',
       'Annual_Premium', 'Annual_Premium_BIL', 'Annual_Premium_PDL',
       'Annual_Premium_PPA', 'Annual_Premium_PI', 'Annual_Premium_ODPE',
       'Annual_Premium_OD', 'Annual_Premium_PE', 'Annual_Premium_FB',
       'Annual_Premium_FBB', 'Annual_Premium_FBP', 'Annual_Premium_LE',
       'Annual_Premium_AXAP', 'Annual_Premium_PSE', 'Annual_Premium_PPD',
       'Annual_Premium_PL', 'Annual_Premium_HK', 'Annual_Premium_PET',
       'Annual_Premium_EQ', 'Payment_Method', 'Payment_Type1',
       'Internet_Discount', 'OD_Type', 'PDL_Limit', 'OD_Coverage', 'Deduct',
       'PPA_Coverage', 'PPA_Limit', 'PI_Coverage', 'PI_Limit', 'PI_Type',
       'FB_Coverage', 'LE_Coverage', 'PDL_Special_Coverage', 'Anshin_Plus',
       'An

In [5]:
lloss.describe()

Event_Number  Claim_Number      Coverage1      Coverage2      PI_Class  \
count  5.180770e+05  4.621880e+05  462188.000000  462188.000000  95185.000000   
mean   9.720327e+06  1.171118e+07       3.116505       3.684801      2.925892   
std    1.248416e+06  1.555914e+06       1.571757       2.453307      0.318921   
min    7.541881e+06  8.960750e+06       1.000000       1.000000      1.000000   
25%    8.616463e+06  1.033523e+07       2.000000       2.000000      3.000000   
50%    9.758577e+06  1.177668e+07       3.000000       3.000000      3.000000   
75%    1.075232e+07  1.305748e+07       4.000000       4.000000      3.000000   
max    1.239077e+07  1.492181e+07      10.000000      19.000000      3.000000   

       IN_OUTPATIENT      Incurred  Contract_Status  Annual_Premium  \
count   74812.000000  5.180770e+05    517770.000000   518077.000000   
mean        2.326940  1.653272e+05         8.630765    56748.640221   
std         1.210362  9.215794e+05         3.952935    32042.557641   
min         1.000000 -1.193140e+06         6.000000     8870.000000   
25%         1.000000  0.000000e+00         7.000000    36470.000000   
50%         3.000000  5.000000e+04         7.000000    49580.000000   
75%         3.000000  1.768420e+05         7.000000    67800.000000   
max         5.000000  2.000270e+08        24.000000   612000.000000   

       Annual_Premium_BIL      ...          EQ_Coverage  License_Color  \
count       518077.000000      ...        518077.000000  518077.000000   
mean          9884.482295      ...             0.013085       2.566518   
std           6030.328908      ...             0.113639       0.524390   
min           2023.000000      ...             0.000000       1.000000   
25%           6420.000000      ...             0.000000       2.000000   
50%           8265.000000      ...             0.000000       3.000000   
75%          11304.000000      ...             0.000000       3.000000   
max         118572.000000      ...             1.000000       3.000000   

             Mileage          Usage       Car_Type        Car_Age  \
count  518077.000000  518077.000000  518077.000000  518077.000000   
mean        1.596533       2.248594       1.967113       5.904676   
std         0.684723       0.468868       0.844692       4.347452   
min         1.000000       1.000000       1.000000       0.000000   
25%         1.000000       2.000000       1.000000       2.000000   
50%         1.000000       2.000000       2.000000       6.000000   
75%         2.000000       3.000000       3.000000       9.000000   
max         3.000000       3.000000       6.000000      80.000000   

       Previous_Accidents    Insured_Sex    Insured_Age  Renewal_Times  
count       518077.000000  518077.000000  518077.000000  518077.000000  
mean             0.078519       1.416633      45.668277       3.290046  
std              0.268987       0.493001      14.557836       3.098196  
min              0.000000       1.000000      17.000000       0.000000  
25%              0.000000       1.000000      35.000000       1.000000  
50%              0.000000       1.000000      42.000000       2.000000  
75%              0.000000       2.000000      56.000000       5.000000  
max              1.000000       2.000000      98.000000      15.000000  

[8 rows x 55 columns]

# missing values
- " Claim_Number " (under Event_Number, an accident can result in multiple claim numbers):   55889 
- " Coverage1 " (coverage type 1) :   55889
- " Coverage2 " (coverage type 2) :   55889
- " PI_Class "  (personal injury class):  422892
- " IN_OUTPATIENT " (whether in the hospital) :  443265
- " Driver " (driver at the time of accident): 40822
- " Contract_Status " (status of the contract):  307
- " Prev_Insurer_Category " (category of previous insurer):  43
- " Original_Channel "  (the method of reachingus for the first time):  44

In [6]:
lloss.isnull().sum()

Event_Number                   0
Claim_Number               55889
Event_Occurrence_Month         0
Inception_Month                0
Issued_Date                    0
Issued_Time                    0
Coverage1                  55889
Coverage2                  55889
PI_Class                  422892
IN_OUTPATIENT             443265
Incurred                       0
EV_LARGE                       0
Driver                     40822
Contract_Status              307
Reason                         0
Effective_Date                 0
Annual_Premium                 0
Annual_Premium_BIL             0
Annual_Premium_PDL             0
Annual_Premium_PPA             0
Annual_Premium_PI              0
Annual_Premium_ODPE            0
Annual_Premium_OD              0
Annual_Premium_PE              0
Annual_Premium_FB              0
Annual_Premium_FBB             0
Annual_Premium_FBP             0
Annual_Premium_LE              0
Annual_Premium_AXAP            0
Annual_Premium_PSE             0
          

# explore target variable "EV_LARGE"

In [7]:
lloss['EV_LARGE'].isnull().sum()

0

In [8]:
# how large each class is 
lloss['EV_LARGE'].value_counts()

EV_000M    466851
EV_001M     32146
EV_002M      9353
EV_004M      5312
EV_003M      2787
EV_010M      1574
EV_090M        54
Name: EV_LARGE, dtype: int64

#EV_OO0M means less than 1M : non-large loss class with 466851 records

#Addition of EV_OO1M to EV_090M : large loss class with 51226 records

#Definitely skewed and unbalanced data

#redefine target varible

- replace 'EV_000M 'with 0, otherwise 1

In [9]:
lloss['EV_LARGE'] = lloss['EV_LARGE'].replace(['EV_001M','EV_002M','EV_004M','EV_003M','EV_010M',
                          'EV_090M'],1)

In [10]:
lloss['EV_LARGE'] = lloss['EV_LARGE'].replace(['EV_000M'],0)

In [11]:
lloss['EV_LARGE'].value_counts()

0    466851
1     51226
Name: EV_LARGE, dtype: int64

# explore target variable "Incurred" for regression

In [12]:
lloss['Incurred'].value_counts()

0          199653
50000       10776
10000        9890
150000       4767
100000       4586
20000         847
200000        778
3150          763
5250          702
3000          676
300000        652
2100          459
4000          353
4200          327
5000          327
450000        269
5400          266
550000        238
1050          227
400000        207
10500         199
8400          191
8000          188
120000        185
140000        177
500000        173
210000        173
160000        169
180000        167
3240          163
            ...  
384293          1
800024          1
48521           1
279830          1
285973          1
816400          1
300300          1
306443          1
791910          1
326913          1
441673          1
435530          1
437579          1
1233342         1
216668          1
54660           1
585091          1
58754           1
3734913         1
468346          1
1426178         1
484722          1
2023982         1
170454          1
400733    

# working on missing values

In [13]:
lloss['PI_Class'].value_counts()

3.0    89706
2.0     3904
1.0     1575
Name: PI_Class, dtype: int64

In [14]:
lloss['PI_Class'] = lloss['PI_Class'].fillna(0)
lloss['IN_OUTPATIENT'] = lloss['IN_OUTPATIENT'].fillna(0)

# delete irrelevant features

- 'Driver'

- 'Coverage1'

- 'Coverage2

In [15]:
lloss = lloss.drop(['Coverage1','Coverage2','Driver'], axis=1)

# clean independent variables

In [16]:
lloss['Reason'] = lloss['Reason'].replace('契約開始', 'The contract begins')
lloss['Reason'] = lloss['Reason'].replace('車両入替', 'Replacement vehicle')
lloss['Reason'] = lloss['Reason'].replace('パートナー情報', 'Partner Information')
lloss['Reason'] = lloss['Reason'].replace('その他の変更', 'Other changes')
lloss['Reason'] = lloss['Reason'].replace('運転者年齢条件', 'Driver age conditions')
lloss['Reason'] = lloss['Reason'].replace('VAN訂正', 'VAN Revised')
lloss['Reason'] = lloss['Reason'].replace('家族限定特約', 'Family limited special')
lloss['Reason'] = lloss['Reason'].replace('車両保険', 'Vehicle insurance')
lloss['Reason'] = lloss['Reason'].replace('使用目的', 'Intended use')
lloss['Reason'] = lloss['Reason'].replace('距離区分', 'Distance classification')
lloss['Reason'] = lloss['Reason'].replace('免許証の色', 'License color')
lloss['Reason'] = lloss['Reason'].replace('居住地域', 'Residential area')
lloss['Reason'] = lloss['Reason'].replace('等級訂正', 'Grade correction')
lloss['Reason'] = lloss['Reason'].replace('ファミリーバイク', 'Family bike')
lloss['Reason'] = lloss['Reason'].replace('登録番号', 'Registration number')
lloss['Reason'] = lloss['Reason'].replace('車台番号', 'Chassis number')
lloss['Reason'] = lloss['Reason'].replace('記名被保険者の変更', 'Changes to the name insured')
lloss['Reason'] = lloss['Reason'].replace('人身傷害補償担保特約', 'Personal injury compensation security agreement')
lloss['Reason'] = lloss['Reason'].replace('誤計上による訂正', 'Correction by miscorrection')
lloss['Reason'] = lloss['Reason'].replace('車両所有者の変更', 'Change of vehicle owner')
lloss['Reason'] = lloss['Reason'].replace('搭乗者傷害保険金額', 'Passenger injury insurance amount')
lloss['Reason'] = lloss['Reason'].replace('ﾍﾟｯﾄﾌﾟﾗｽ', 'Pet Plus')
lloss['Reason'] = lloss['Reason'].replace('弁護士費用(人+物)', 'Attorney fee (person + thing)')
lloss['Reason'] = lloss['Reason'].replace('免責金額の変更', 'Change of deductible amount')
lloss['Reason'] = lloss['Reason'].replace('自動継続特約', 'Automatic continuation rider')
lloss['Reason'] = lloss['Reason'].replace('対物賠償金額', 'Object reparation amount')
lloss['Reason'] = lloss['Reason'].replace('ﾌｧﾐﾘｰﾌﾟﾗｽ', 'Family Plus')
lloss['Reason'] = lloss['Reason'].replace('車両免責ゼロ特約', 'Vehicle disclaimer Zero rider')
lloss['Reason'] = lloss['Reason'].replace('対物全損時費用', 'Objective total loss cost')
lloss['Reason'] = lloss['Reason'].replace('口座情報', 'Account information')
lloss['Reason'] = lloss['Reason'].replace('任意解約', 'Arbitrary cancellation')
lloss['Reason'] = lloss['Reason'].replace('未払い解除', 'Unpaid released')
lloss['Reason'] = lloss['Reason'].replace('車両価額', 'Vehicle price')
lloss['Reason'] = lloss['Reason'].replace('ﾚﾃﾞｨｰｽﾌﾟﾗｽ', "Women's Plus")
lloss['Reason'] = lloss['Reason'].replace('地震全損一時金', 'Earthquake total loss lump sum')
lloss['Reason'] = lloss['Reason'].replace('初度登録年月', 'First registration date')
lloss['Reason'] = lloss['Reason'].replace('エアバッグ', 'Airbag')
lloss['Reason'] = lloss['Reason'].replace('免許取得日', 'License Acquisition Date')
lloss['Reason'] = lloss['Reason'].replace('約款違反解除（日割）', 'Cancellation of violation of contracts (daily)')
lloss['Reason'] = lloss['Reason'].replace('任意解約(会社都合)', 'Optional cancellation (company convenience)')
lloss['Reason'] = lloss['Reason'].replace('契約取消', 'Contract cancellation')

In [17]:
import re
lloss['Reason'] = lloss['Reason'].apply(lambda x: re.sub(' +',' ',x.lower()))

# feature: 'Annual_Premium'

In [18]:
lloss['Annual_Premium'].value_counts().sort_values()

357600      1
167110      1
231720      1
202420      1
124650      1
183040      1
111820      1
201880      1
174210      1
11680       1
142620      1
150080      1
225940      1
197530      1
358200      1
239330      1
166930      1
144060      1
152240      1
113120      1
124330      1
264840      1
207980      1
117230      1
135140      1
160180      1
187020      1
148190      1
110140      1
203040      1
         ... 
48720     343
52560     344
43200     345
56520     346
47640     346
51840     346
41400     347
46680     348
46200     348
52800     351
48480     351
46320     351
49560     352
48360     352
46920     353
45840     353
43560     354
44640     355
45240     355
41520     355
46800     356
46080     359
41280     362
44400     365
47880     368
40920     368
55680     369
53640     376
49800     377
48000     398
Name: Annual_Premium, dtype: int64

In [19]:
lloss['Annual_Premium'].max()

612000

In [20]:
lloss['Annual_Premium'].min()

8870

#['Annual_Premium'] : range from 8870 to 612000

In [22]:
lloss['Annual_Premium_BIL'].value_counts().sort_values()

75648      1
21403      1
21491      1
23538      1
23450      1
35660      1
17817      1
23047      1
25671      1
19352      1
14859      1
2573       1
17063      1
33613      1
21569      1
21702      1
17156      1
39768      1
3287       1
17305      1
20721      1
16338      1
29599      1
24839      1
3064       1
25491      1
27554      1
43652      1
17493      1
41652      1
        ... 
7200     218
9000     218
7764     218
9240     219
6888     219
7092     220
6900     222
8400     223
9252     224
6420     224
7992     224
7512     225
6636     225
9312     226
7872     230
6924     231
7272     231
6492     231
8028     231
7416     233
8304     234
6912     235
6552     238
7824     239
7728     240
7440     240
8976     244
6780     250
8136     252
7548     261
Name: Annual_Premium_BIL, dtype: int64

In [23]:
lloss['Annual_Premium_PDL'].value_counts().sort_values()

6147        1
83796       1
42856       1
30578       1
89316       1
34015       1
6014        1
50391       1
6270        1
106825      1
33133       1
29889       1
6526        1
45675       1
70495       1
43880       1
44649       1
63074       1
36318       1
32115       1
37501       1
25443       1
30331       1
38759       1
56676       1
74328       1
26563       1
35437       1
29298       1
98016       1
         ... 
12360     142
13968     142
14208     143
13680     143
14136     144
13104     144
13488     145
15204     145
13416     146
14064     146
13896     146
11688     146
13536     146
13860     147
15084     147
13740     148
14268     148
16044     148
13956     148
13140     148
14544     149
15288     149
12612     150
14388     150
13308     151
13980     152
14196     155
14124     156
12288     157
14436     170
Name: Annual_Premium_PDL, dtype: int64

In [24]:
lloss['Annual_Premium_PPA'].value_counts().sort_values()

8934         1
11964        1
6614         1
5563         1
8688         1
5682         1
6909         1
7925         1
9660         1
6391         1
9880         1
6707         1
5329         1
7376         1
9768         1
6863         1
13056        1
6607         1
5927         1
6181         1
5939         1
6070         1
6193         1
7172         1
6583         1
5588         1
6065         1
6449         1
6326         1
8380         1
         ...  
1692       872
1656       875
1416       877
1464       878
1308       879
1368       882
1596       887
1680       892
1344       893
1704       893
1356       893
1536       896
1476       897
1500       901
1548       907
1512       913
1728       913
1332       913
1428       914
1488       915
1608       922
1848       922
1392       925
1740       926
1404       934
1524       938
1320       950
1572       961
1560       975
0        27662
Name: Annual_Premium_PPA, dtype: int64

In [25]:
lloss['Annual_Premium_PI'].value_counts().sort_values()

13013        1
13233        1
14417        1
22367        1
22934        1
15294        1
13626        1
17841        1
11454        1
10589        1
1947         1
14396        1
13245        1
19378        1
1879         1
12698        1
11186        1
10907        1
23484        1
11698        1
12735        1
16319        1
12859        1
1976         1
13115        1
19860        1
13627        1
14780        1
13246        1
17328        1
         ...  
4308       333
3972       334
4752       334
4380       335
4536       338
4896       338
5436       338
4044       338
4812       340
4704       341
4572       345
5232       348
4008       351
5040       353
3840       354
4236       355
5112       355
4080       361
4440       361
4272       361
4332       361
4776       368
3984       368
5064       369
4848       370
4824       383
4668       396
4188       396
4608       396
0        89207
Name: Annual_Premium_PI, dtype: int64

In [26]:
lloss['Annual_Premium_ODPE'].value_counts().sort_values()

8188           1
36057          1
37769          1
38106          1
12565          1
56607          1
96948          1
45965          1
53513          1
41888          1
14612          1
39847          1
43270          1
35979          1
68766          1
61709          1
131155         1
66452          1
58280          1
48403          1
145884         1
47598          1
80316          1
70890          1
45453          1
33163          1
62625          1
47620          1
43526          1
59782          1
           ...  
24612         61
19632         61
18000         61
20796         61
21684         61
25500         61
18216         62
20844         62
23712         63
16068         63
20820         64
30576         65
19176         65
23376         65
23520         65
18720         65
23904         65
21780         65
17016         65
18300         66
20508         66
28824         67
23832         67
29580         68
21120         68
14724         69
20904         71
22776         

In [27]:
lloss['Annual_Premium_OD'].value_counts().sort_values()

109898         1
44694          1
56656          1
56984          1
112309         1
48476          1
47763          1
58527          1
89440          1
155136         1
45714          1
84397          1
53662          1
51613          1
43409          1
71017          1
41360          1
36958          1
130236         1
44894          1
9863           1
7808           1
67435          1
73120          1
50587          1
72864          1
52634          1
46485          1
10313          1
6223           1
           ...  
14868         62
26640         63
17364         63
23160         63
20940         63
22008         64
17688         64
19476         64
28440         65
24672         65
16512         65
21996         65
16884         66
23580         66
21252         66
23328         67
23292         68
22344         70
27444         70
16200         70
23304         70
23592         72
23832         72
20352         73
18684         75
26076         75
21060         76
20160         

In [28]:
lloss['Annual_Premium_OD'].max()

506076

#"Annual_Premium_OD"
- definition of "OD" : damage to your own vehicle
- numerical data ranging from "0" to "506076" because of different premium policies
- value "0" appears most : 151642 times
- "0" means "This customer didn't pay annual premium for car damage"?

In [29]:
lloss['Annual_Premium_PE'].value_counts().sort_values()

1485         1
1726         1
2360         1
1593         1
2083         1
2748         1
1508         1
1486         1
1599         1
2868         1
1793         1
1701         1
2033         1
1901         1
103          1
1784         1
1551         1
1925         1
1790         1
1397         1
1847         1
1473         1
1729         1
1326         1
2048         1
1837         1
2484         1
1994         1
2856         1
1565         1
         ...  
624       1532
288       1611
612       1619
600       1620
588       1692
576       1698
564       1763
300       1768
312       1839
552       1891
324       1903
528       1915
540       1947
336       2042
516       2115
504       2152
432       2276
492       2280
348       2293
444       2296
468       2328
384       2352
360       2359
420       2371
456       2394
396       2394
372       2406
408       2419
480       2436
0       151642
Name: Annual_Premium_PE, dtype: int64

In [30]:
lloss['Annual_Premium_PE'].max()

3780

#"Annual_Premium_PE
- definition: personal effects mean privately owned items
- numerical data ranging from "0" to "3780"
- value "0" appears most: 151642 times

In [31]:
lloss['Annual_Premium_FB'].value_counts().sort_values()

7848        39
6648        48
7692        51
7836        56
7680        75
6576        91
5220       100
8664       132
7475       147
8652       158
7330       166
5172       197
6333       247
7319       250
7464       258
8258       297
6268       312
8246       442
4970       508
4919       598
7728      1263
7056      1321
7884      1392
5544      2183
7361      3707
8700      3883
6725      3910
7507      4062
5277      7329
8294      9759
0       475096
Name: Annual_Premium_FB, dtype: int64

In [32]:
lloss['Annual_Premium_FB'].max()

8700

#"Annual_Premium_FB" 
- means family bike = motorcycle 
- numerical data ranging from "0" to "8700"

In [33]:
lloss['Annual_Premium_FBP'].value_counts().sort_values()

3624        39
3108        48
3552        51
3612        56
3540        75
3036        91
2436       100
3996       132
3448       147
3984       158
3381       166
2388       197
2957       247
3370       250
3437       258
3809       297
2892       312
3797       442
2321       508
2270       598
3588      1263
3516      1321
3660      1392
2760      2183
3412      3707
4032      3883
3349      3910
3480      4062
2628      7329
3845      9759
0       475096
Name: Annual_Premium_FBP, dtype: int64

In [34]:
lloss['Annual_Premium_FBP'].max()

4032

#the result is the same as the one for "Annual_Premium_FB"

" Annual_Premium_FBP"

- definition: family bike property damage

- maximum : 4032

In [35]:
lloss['Annual_Premium_FBB'].value_counts().sort_values()

4140      1389
3540      1460
4224      1487
2784      2480
3949      4123
4668      4173
4027      4467
3376      4469
2649      8435
4449     10498
0       475096
Name: Annual_Premium_FBB, dtype: int64

In [36]:
lloss['Annual_Premium_FBB'].max()

4668

In [37]:
lloss['Annual_Premium_LE'].value_counts().sort_values()

3096       181
2947       645
3756       899
3573      2156
3144      4166
3408     12788
3084     13495
2994     14253
3248     35632
2935     46504
0       387358
Name: Annual_Premium_LE, dtype: int64

In [38]:
lloss['Annual_Premium_LE'].max()

3756

In [39]:
lloss['Annual_Premium_AXAP'].value_counts().sort_values()

2821         1
2766         1
2976         1
3046         1
1272         1
1224         2
2841         2
3120         2
1200         2
2952         3
1168         3
2888         4
1207         4
2856         5
3072         5
1145         6
3118         6
3180         8
2660         9
2964        10
1308        10
3252        10
2804        12
2713        12
3300        15
3228        18
2724        19
1248        20
2928        21
2594        21
         ...  
993        202
1281       206
2857       220
973        237
3036       301
2868       337
2970       394
2859       416
1028       420
1116       429
2632       439
924        447
2684       513
969        609
1092       633
1068       635
950        669
1080       944
2898      1036
1073      1081
879       1328
2736      1361
1164      1640
1048      1750
1152      1892
1011      2481
1031      2874
1114      6360
1102      7810
0       477300
Name: Annual_Premium_AXAP, dtype: int64

In [40]:
lloss['Annual_Premium_AXAP'].max()

3300

#"Annual_Premium_AXAP"

- definition: Anshin plus
    
- what does "Anshin plus" mean?

- "0" appear most

In [41]:
lloss['Annual_Premium_PSE'].value_counts().sort_values()

216       171
129       274
132       349
205       520
144       779
142      1382
180      1938
192      2227
288      2546
167      3142
170      3759
183      8040
279      9867
0      483083
Name: Annual_Premium_PSE, dtype: int64

In [42]:
lloss['Annual_Premium_PSE'].max()

288

In [43]:
lloss['Annual_Premium_PPD'].value_counts().sort_values()

1776       244
1812       246
1848       600
1687       641
1944       733
1720       826
1856      1825
1757      1947
0       511015
Name: Annual_Premium_PPD, dtype: int64

In [44]:
lloss['Annual_Premium_PPD'].max()

1944

#"Annual_Premium_PPD"

- almost "0"

In [45]:
lloss['Annual_Premium_PL'].value_counts().sort_values()

816      1137
828      1380
732      3164
900      3504
778      4085
794      4620
859     11172
700     11715
0      477300
Name: Annual_Premium_PL, dtype: int64

In [46]:
lloss['Annual_Premium_PL'].max()

900

In [47]:
lloss['Annual_Premium_HK'].value_counts().sort_values()

132      2334
66       2943
67       3447
123      8932
72      12370
0      488051
Name: Annual_Premium_HK, dtype: int64

In [48]:
lloss['Annual_Premium_HK'].max()

132

In [49]:
lloss['Annual_Premium_PE'].value_counts().sort_values()

1485         1
1726         1
2360         1
1593         1
2083         1
2748         1
1508         1
1486         1
1599         1
2868         1
1793         1
1701         1
2033         1
1901         1
103          1
1784         1
1551         1
1925         1
1790         1
1397         1
1847         1
1473         1
1729         1
1326         1
2048         1
1837         1
2484         1
1994         1
2856         1
1565         1
         ...  
624       1532
288       1611
612       1619
600       1620
588       1692
576       1698
564       1763
300       1768
312       1839
552       1891
324       1903
528       1915
540       1947
336       2042
516       2115
504       2152
432       2276
492       2280
348       2293
444       2296
468       2328
384       2352
360       2359
420       2371
456       2394
396       2394
372       2406
408       2419
480       2436
0       151642
Name: Annual_Premium_PE, dtype: int64

In [50]:
lloss['Annual_Premium_PE'].max()

3780

In [51]:
lloss['Annual_Premium_EQ'].value_counts().sort_values()

1572         5
1500        15
2100        36
3156        48
4728        52
2500        52
2000        65
3000        73
3672        78
2628        80
3500        95
4000        97
4200       110
4500       132
5256      1726
5000      4115
0       511298
Name: Annual_Premium_EQ, dtype: int64

In [52]:
lloss['Annual_Premium_EQ'].max()

5256

#recap:
- ['Annual_Premium_BIL'] (bodily injury liability): no value '0'    
- ['Annual_Premium_PDL'] (property damage liability): no value '0'
- others, '0' appears most frequently

# feature: 'Coverage'

In [53]:
#own damage 
lloss['OD_Coverage'].value_counts().sort_values()

0    151646
1    366431
Name: OD_Coverage, dtype: int64

In [54]:
#passenger's personal accident
lloss['PPA_Coverage'].value_counts().sort_values()

0     27662
1    490415
Name: PPA_Coverage, dtype: int64

In [55]:
#personal injury
lloss['PI_Coverage'].value_counts().sort_values()

0     89207
1    428870
Name: PI_Coverage, dtype: int64

In [56]:
#family bike
lloss['FB_Coverage'].value_counts().sort_values()

1     42981
0    475096
Name: FB_Coverage, dtype: int64

In [57]:
# lawyer expense
lloss['LE_Coverage'].value_counts().sort_values()

1    130719
0    387358
Name: LE_Coverage, dtype: int64

In [58]:
#property damage specical
lloss['PDL_Special_Coverage'].value_counts().sort_values()

1     85835
0    432242
Name: PDL_Special_Coverage, dtype: int64

In [59]:
# earthquake option
lloss['EQ_Coverage'].value_counts().sort_values()

1      6779
0    511298
Name: EQ_Coverage, dtype: int64

# feature: 'Anshin_Plus' related

In [60]:
lloss['Anshin_Plus'].value_counts().sort_values()

1     40777
0    477300
Name: Anshin_Plus, dtype: int64

In [61]:
lloss['Anshin_Plus_Ladies'].value_counts().sort_values()

1      7062
0    511015
Name: Anshin_Plus_Ladies, dtype: int64

In [62]:
lloss['Anshin_Plus_Pet'].value_counts().sort_values()

1      7061
0    511016
Name: Anshin_Plus_Pet, dtype: int64

In [63]:
lloss['Anshin_Plus_Family'].value_counts().sort_values()

1     30026
0    488051
Name: Anshin_Plus_Family, dtype: int64

# Car related

In [64]:
lloss['Car_Type'].value_counts().sort_values()

6        53
4      2441
5      7993
3    120775
1    166886
2    219929
Name: Car_Type, dtype: int64

In [65]:
lloss['Car_Name'].value_counts().sort_values()

VH                 1
SW2H41             1
ｸﾗｳﾝｺﾝﾌｵ-ﾄ         1
YR25V              1
LN61V              1
ﾌﾟﾗｳﾃﾞｲｱ           1
KR2                1
DB234IK            1
BF5V               1
L70S               1
NCP122             1
SJ30               1
L70V               1
YY61               1
PJC22              1
DB71T              1
PZJ77V             1
ﾃﾙｽﾀ-2             1
ﾋﾟｱﾂﾂｱ             1
EE107V             1
X-90               1
LH85               1
ﾗﾝﾁｱ               1
ﾌﾟﾚﾂｿ              1
ﾌﾟﾛｼ-ﾄﾞ            1
NT100ｸﾘﾂﾊﾟ-        1
DA71V              1
PG6SS              1
KR27V              1
CR36V              1
               ...  
ｸﾗｳﾝ            4353
ﾓｺ              4479
ｽｲﾌﾄ            4764
ｽﾄﾘ-ﾑ           4783
ﾊﾟﾂｿ            4832
ﾚｶﾞｼｲ           5568
ｼｴﾝﾀ            5595
BB              5786
ｳｲﾂｼﾕ           5808
ﾉ-ﾄ             5988
ﾒﾙｾﾃﾞｽﾍﾞﾝﾂ      6203
ｵﾃﾞﾂｾｲ          6576
ﾌﾘ-ﾄﾞ           6961
ﾏ-ﾁ             7234
ﾌｵﾙｸｽﾜ-ｹﾞﾝ      8004
BMW             8016
ﾃﾞﾐｵ         

In [66]:
lloss['Car_Age'].value_counts().sort_values()

45        1
77        1
68        1
34        1
43        1
35        1
55        1
41        2
36        2
52        2
80        2
69        2
40        3
67        3
44        3
39        3
42        3
71        3
31        4
26        4
32        4
70        4
46        4
75        5
72        5
74        6
76        6
73        7
30        7
37       11
27       11
29       11
28       17
25       29
24       59
23      117
22      228
21      392
20      510
19      969
18     1642
17     2523
16     4204
15     6246
14     9450
13    13311
12    17578
11    22099
10    27032
9     32108
8     37368
2     38840
1     39553
3     39858
7     40826
4     41629
5     41701
6     42879
0     56785
Name: Car_Age, dtype: int64

# Insured_related

In [67]:
lloss['Insured_Sex'].value_counts().sort_values()

2    215848
1    302229
Name: Insured_Sex, dtype: int64

In [68]:
lloss['Insured_Age'].value_counts().sort_values()

98        1
96        1
93        2
94        3
92       12
91       16
90       31
89       66
17      122
88      126
87      152
86      235
85      337
84      486
83      645
82      864
81     1117
80     1306
79     1542
18     1635
19     1793
78     2011
20     2108
77     2381
21     2555
76     2668
75     3106
22     3127
74     3505
23     3865
      ...  
54     7703
64     7719
62     7874
53     8081
63     8310
52     8414
51     8428
50     8692
28     8732
49     9054
48     9295
47     9648
46    10449
45    10508
29    10549
44    11507
30    11755
43    12682
31    12789
42    13620
41    14345
32    14354
40    15465
33    15715
39    16667
34    16860
35    17354
38    17552
36    17752
37    17872
Name: Insured_Age, dtype: int64

In [69]:
lloss['Insured_Age_Range'].value_counts().sort_values()

00-19      3550
20-29     54013
50-59     76923
60-      108348
40-49    116573
30-39    158670
Name: Insured_Age_Range, dtype: int64

In [70]:
lloss['Region'].value_counts().sort_values()

(06) 四国        6631
(01) 北海道      10878
(07) 九州       53479
(04) 北陸東海     55961
(02) 東北       59784
(05) 近畿中国     66936
(03) 関東      264408
Name: Region, dtype: int64

In [71]:
lloss['Prefecture'].value_counts().sort_values()

島根県       786
鳥取県       962
高知県      1097
沖縄県      1211
徳島県      1267
和歌山県     1737
福井県      2057
香川県      2109
愛媛県      2158
山口県      2404
佐賀県      2410
富山県      2667
宮崎県      2668
石川県      3389
滋賀県      3501
大分県      3670
奈良県      3689
長崎県      3816
鹿児島県     3904
岡山県      3968
山梨県      3979
秋田県      4981
岐阜県      5090
三重県      5250
京都府      5928
青森県      6218
長野県      6521
岩手県      6663
新潟県      6665
広島県      6829
熊本県      6886
山形県      8253
栃木県      8973
群馬県      9311
北海道     10878
静岡県     12248
福島県     13157
兵庫県     15662
茨城県     15792
宮城県     20512
大阪府     21470
愛知県     25260
福岡県     28914
千葉県     48021
埼玉県     52336
東京都     55527
神奈川県    57283
Name: Prefecture, dtype: int64

# other features

In [72]:
lloss['Payment_Method'].value_counts().sort_values()

9        1958
11       2307
10      12086
999     81956
3      168804
1      250966
Name: Payment_Method, dtype: int64

#definition:

- "1": "credit card"

- "3": "Offline CVS (convenience store)"
    
- "9": "NTT-if"

- "10": "Online CVS"
    
- "11": "Net-Bank"
    
- "999": "unknown"

In [73]:
lloss['Payment_Type1'].value_counts().sort_values()

7    126490
5    391587
Name: Payment_Type1, dtype: int64

#definition:

- "5": "paid full at once"
    
- "7": "installment"

In [74]:
lloss['Internet_Discount'].value_counts().sort_values()

1    256707
0    261370
Name: Internet_Discount, dtype: int64

In [75]:
#own damage type - ['OD_Coverage']
lloss['OD_Type'].value_counts().sort_values()

2     77375
1    151642
3    289060
Name: OD_Type, dtype: int64

In [76]:
# property damage coverage limit
lloss['PDL_Limit'].value_counts().sort_values()

(04) 20m           21790
(03) 10m           38125
(01) Unlimited    458162
Name: PDL_Limit, dtype: int64

In [77]:
#passenger's personal accident coverage limit
lloss['PPA_Limit'].value_counts().sort_values()

(06) 20m     23213
(00) NO      27662
(05) 15m     38505
(03) 5m      82261
(04) 10m    346436
Name: PPA_Limit, dtype: int64

In [78]:
# personal injury coverage limit
lloss['PI_Limit'].value_counts().sort_values()

(08) 90m              78
(05) 60m             529
(03) 40m            1400
(06) 70m            6036
(01) Unlimited      7160
(09) 100m          26996
(07) 80m           27942
(00) NO            89207
(04) 50m          160182
(02) 30m          198547
Name: PI_Limit, dtype: int64

In [79]:
# personal injury type
lloss['PI_Type'].value_counts().sort_values()

2     32590
1     89207
3    396280
Name: PI_Type, dtype: int64

In [80]:
lloss['Deduct'].value_counts().sort_values()

(07) 30% Deductible       237
4                        2964
5                       12358
2                      110773
6                      119940
3                      120159
1                      151646
Name: Deduct, dtype: int64

In [81]:
lloss['Automatic_Renewal'].value_counts().sort_values()

0      3728
1    514349
Name: Automatic_Renewal, dtype: int64

In [82]:
lloss['Age_Restriction'].value_counts().sort_values()

5      4432
1     22530
2     46430
3     62999
4    381686
Name: Age_Restriction, dtype: int64

In [83]:
lloss['Family_Restriction'].value_counts().sort_values()

3     42823
2     66435
0    146856
1    261963
Name: Family_Restriction, dtype: int64

In [84]:
lloss['Tariff'].value_counts().sort_values()

(01) AUTO    518077
Name: Tariff, dtype: int64

In [85]:
lloss['Bonus_Malus'].value_counts().sort_values()

1           435
2          1251
3          4098
4          6417
5          7823
07New      8405
06F       13341
19        18733
16        22145
18        22540
15        24492
14        25896
17        26135
06New     26541
13        26874
07F       26979
12        27807
11        28744
10        29180
9         29655
8         31137
20       109449
Name: Bonus_Malus, dtype: int64

In [86]:
lloss['Bonus_Malus_Range'].value_counts().sort_values()

01_05        20024
06_07New     34946
06_09       101112
20          109449
15_19       114045
10_14       138501
Name: Bonus_Malus_Range, dtype: int64

In [87]:
lloss['License_Color'].value_counts().sort_values()

1      7618
2    209341
3    301118
Name: License_Color, dtype: int64

In [88]:
lloss['Mileage'].value_counts().sort_values()

3     59103
2    190844
1    268130
Name: Mileage, dtype: int64

In [89]:
lloss['Usage'].value_counts().sort_values()

1      8559
3    137350
2    372168
Name: Usage, dtype: int64

In [90]:
lloss['Previous_Accidents'].value_counts().sort_values()

1     40679
0    477398
Name: Previous_Accidents, dtype: int64

In [91]:
lloss['Prev_Insurer_Category'].value_counts().sort_values()

(01) Direct        117928
(99) New           151298
(02) Non Direct    248808
Name: Prev_Insurer_Category, dtype: int64

In [92]:
lloss['Original_Channel'].value_counts().sort_values()

Database marketing        28
(08) Banc                774
(04) Fax                1598
(09) MGM                5128
(07) Agent             27928
(06) Axa Life          51728
(05) Re-quote          52695
(03) Broker            80688
(02) Internet          97626
(01) Phone            199840
Name: Original_Channel, dtype: int64

In [93]:
lloss['Renewal_Times'].value_counts().sort_values()

15         3
14       522
13      2095
12      3941
11      7228
10     10938
9      15038
8      21129
7      27079
6      32217
5      39192
4      46163
3      53242
2      63546
1      79070
0     116674
Name: Renewal_Times, dtype: int64

# tranforming datatype

In [94]:
lloss.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 518077 entries, 0 to 518076
Data columns (total 73 columns):
Event_Number              518077 non-null int64
Claim_Number              462188 non-null float64
Event_Occurrence_Month    518077 non-null object
Inception_Month           518077 non-null object
Issued_Date               518077 non-null object
Issued_Time               518077 non-null object
PI_Class                  518077 non-null float64
IN_OUTPATIENT             518077 non-null float64
Incurred                  518077 non-null int64
EV_LARGE                  518077 non-null int64
Contract_Status           517770 non-null float64
Reason                    518077 non-null object
Effective_Date            518077 non-null object
Annual_Premium            518077 non-null int64
Annual_Premium_BIL        518077 non-null int64
Annual_Premium_PDL        518077 non-null int64
Annual_Premium_PPA        518077 non-null int64
Annual_Premium_PI         518077 non-null int64
Annual_Prem

#Data type for each variable
- float64: 4 variables
    
- int64: 50 variables

- object: 19 variables

so we'd better to transform datatype for following features:

- 'PDL_Limit', 'Deduct', 'PPA_Limit', 'PI_Limit','Tariff','Bonus_Malus','Bonus_Malus_Range','Prev_Insurer_Category',

- 'Original_Channel'

- 'Insured_Age_Range' (unsolved)

text data

- 'Region'                  

- 'Prefecture' 

- 'Car_Name '

- 'Reason'

Datetime
- 'Event_Occurrence_Month'

- 'Inception_Month'

- 'Issued_Date'

- 'Issued_Time'

- 'Effective_Date'


In [95]:
lloss['PDL_Limit'] = lloss['PDL_Limit'].replace('(01) Unlimited', 1)
lloss['PDL_Limit'] = lloss['PDL_Limit'].replace('(03) 10m', 3)
lloss['PDL_Limit'] = lloss['PDL_Limit'].replace('(04) 20m', 4)

In [96]:
lloss['Deduct'] = lloss['Deduct'].replace('(07) 30% Deductible','7')

In [97]:
lloss['Deduct'] = lloss['Deduct'].astype(int)

In [98]:
def transform_dtype(feature,old_value,new_value):
    lloss[feature] = lloss[feature].replace(old_value,new_value)
    return type(lloss[feature].iloc[0])

In [99]:
transform_dtype('PPA_Limit',['(04) 10m','(03) 5m','(05) 15m','(00) NO','(06) 20m'], [4,3,5,0,6])

numpy.int64

In [100]:
transform_dtype('PI_Limit',['(02) 30m','(04) 50m','(00) NO','(07) 80m','(09) 100m','(01) Unlimited','(06) 70m',
                           '(03) 40m','(05) 60m','(08) 90m'], [2,4,0,7,9,1,6,3,5,8])

numpy.int64

In [101]:
transform_dtype('Tariff','(01) AUTO', 1)

numpy.int64

In [102]:
transform_dtype('Bonus_Malus',['07F','06New','06F','07New'], [7,60,6,70])

str

In [103]:
lloss['Bonus_Malus'] = lloss['Bonus_Malus'].astype(int)

In [104]:
transform_dtype('Bonus_Malus_Range','06_07New','06_07')

str

In [105]:
lloss['Bonus_Malus_Range'] = lloss['Bonus_Malus_Range'].astype(int)

In [106]:
# fillna in ['Prev_Insurer_Category']
lloss['Prev_Insurer_Category'] = lloss['Prev_Insurer_Category'].fillna(0)

In [107]:
transform_dtype('Prev_Insurer_Category',['(01) Direct','(99) New','(02) Non Direct'],[1,99,2])

numpy.int64

In [108]:
# fillna in ['Original_Channel']
lloss['Original_Channel'] = lloss['Original_Channel'].fillna(0)

In [109]:
transform_dtype('Original_Channel',['(01) Phone','(02) Internet','(03) Broker','(05) Re-quote',
                    '(06) Axa Life','(07) Agent','(09) MGM','(04) Fax','(08) Banc','Database marketing'],
                [1,2,3,5,6,7,9,4,8,10])

numpy.int64

# working on timestamp

#breaking datatime columns : break into 'year', 'month', and 'date'
- 'Event_Occurrence_Month'  

- 'Inception_Month'           

- 'Issued_Date'              

- 'Issued_Time'               

- 'Effective_Date' 

In [110]:
def datetime(new_feature,feature,year,month,date):
    lloss[new_feature] = pd.to_datetime(lloss[feature])
    lloss[year] = lloss[new_feature].dt.year
    lloss[month] = lloss[new_feature].dt.month
    lloss[date] = lloss[new_feature].dt.date
    return lloss[year].head()

In [111]:
datetime('Event_Occurrence','Event_Occurrence_Month','Event_Occurrence_year','Event_Occurrence_month','Event_Occurrence_date')
datetime('Inception','Inception_Month','Inception_year', 'Inception_month','Inception_date')
datetime('Issued','Issued_Date','Issued_year','Issued_month','Issued_date')
datetime('Effective','Effective_Date','Effective_year','Effective_month','Effective_date')

0    2010
1    2010
2    2010
3    2010
4    2010
Name: Effective_year, dtype: int64

In [119]:
columns = list(lloss.columns)
new_claimdata_1 = lloss[columns]
new_claimdata_1.to_csv('new_claimdata_1.csv',index = False)

#so far left tasks:

- "missing observations in ['Claim_Number'],['Contract_Status']"

- whether the way of processing missing values is meaningful and effective?

- text data

- 0 means 'This customer didn't pay annual premium for that category?

- "what does 'Anshin plus' mean for ['Annual_Premium_AXAP'] ?

- why the contract was finalized after the insurance policy has been effective ?

- continue to transform datatype : 'Insured_Age_Range'

In [123]:
# fetures_to_use will be final features used for prediction
# features_to_use = [col for col in columns if col not in ('EV_LARGE','Issued_Date','Effective_Date',
#                                                                  'Inception_Month','Event_Occurrence_Month')]
